File: CallkitSample/CallkitSample/StringeeImplement/CallingViewController.m
//
//  SPCallingViewController.m
//  Softphone
//
//  Created by Hoang Duoc on 7/11/17.
//  Copyright © 2017 Hoang Duoc. All rights reserved.
//

#import "CallingViewController.h"
#import "StringeeImplement.h"
#import "InstanceManager.h"

static int CALL_TIME_OUT = 60; // giây

@interface CallingViewController ()

@end

@implementation CallingViewController {
    NSTimer *timer;
    NSTimer *timeoutTimer;
    int interval;
    
    BOOL _isIncoming;
    BOOL _isVideo;
    BOOL _isMute;
    BOOL _isSpeaker;
    BOOL _localVideoEnabled;
    BOOL _ended;
    SignalingState _signalingState;
    MediaState _mediaState;
    NSString *_from;
    NSString *_to;
    
    BOOL _mediaFirstTimeConnected;
}

- (instancetype)initFrom:(NSString *)from to:(NSString *)to isVideo:(BOOL)isVideo
{
    self = [super initWithNibName:@"CallingViewController" bundle:nil];
    if (self) {
        _isIncoming = false;
        _isVideo = isVideo;
        _from = from;
        _to = to;
        [self commonInit];
    }
    
    return self;
}

- (instancetype)initWithCall:(StringeeCall *)call
{
    self = [super initWithNibName:@"CallingViewController" bundle:nil];
    if (self) {
        self.stringeeCall = call;
        call.delegate = self;
        _isIncoming = true;
        _isVideo = call.isVideoCall;
        _from = call.from;
        _to = call.to;
        [self commonInit];
    }
    
    return self;
}

- (void)commonInit {
    self.modalPresentationStyle = UIModalPresentationFullScreen;
    [InstanceManager instance].callingVC = self;
    _localVideoEnabled = true;
    _signalingState = SignalingStateCalling;
    _mediaState = MediaStateDisconnected;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [[UIDevice currentDevice] setProximityMonitoringEnabled:_isVideo];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(handleAudioSessionRouteChange) name:AVAudioSessionRouteChangeNotification object:nil];
    
    [self startTimeoutTimer];
    [self setupUI];
    
    if (!_isIncoming) {
        // make call
        self.stringeeCall = [[StringeeCall alloc] initWithStringeeClient:[StringeeImplement instance].stringeeClient from:_from to:_to];
        self.stringeeCall.isVideoCall = _isVideo;
        self.stringeeCall.delegate = self;
        [self.stringeeCall makeCallWithCompletionHandler:^(BOOL status, int code, NSString *message, NSString *data) {
            if (status) {
                // Sucess => show callkit screen
                [[CallManager instance] startCallWithPhone:_to calleeName:_to isVideo:_isVideo stringeeCall:self.stringeeCall];
            } else {
                [self endCallAndDismissWithTitle:nil];
            }
        }];
    }
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self name:AVAudioSessionRouteChangeNotification object:nil];
}

// MARK: -  Public

- (void)changeStateToAnswered {
    _signalingState = SignalingStateAnswered;
    [self updateScreen];
}

- (BOOL)getMuteState {
    return _isMute;
}

- (void)setMuteState:(BOOL)mute {
    _isMute = mute;
}

- (void)endCallAndDismissWithTitle:(NSString *)title {
    dispatch_async(dispatch_get_main_queue(), ^{
        if (_ended) {
            return;
        }
        _ended = !_ended;
        
        self.lbStatus.text = title != nil ? title : @"Kết thúc cuộc gọi";
        self.blurView.alpha = 0.4;
        self.view.userInteractionEnabled = NO;
        [[UIDevice currentDevice] setProximityMonitoringEnabled:NO];
        
        [self stopTimeoutTimer];
        [self stopTimer];
        
        [[CallManager instance] endCall];
                
        UIViewController *vc = self.presentingViewController;
        while (vc.presentingViewController) {
            vc = vc.presentingViewController;
        }
        [vc dismissViewControllerAnimated:YES completion:^{
            [InstanceManager instance].callingVC = nil;
        }];
            
    });
    
}

// MARK: -  UI

- (void)setupUI {
    if (_isIncoming) {
        self.lbUsername.text = self.stringeeCall.fromAlias.length > 0 ? self.stringeeCall.fromAlias : self.stringeeCall.from;
    } else {
        self.lbUsername.text = self.stringeeCall.toAlias.length > 0 ? self.stringeeCall.toAlias : self.stringeeCall.to;
    }
    
    [self updateScreen];
}

- (void)updateScreen {
    CallScreenType screenType = [self getScreenType];
    switch (screenType) {
        case CallScreenTypeOutgoingVoice:
            [self changeStackViewButton:self.btAccept hide:true];
            [self changeStackViewButton:self.btReject hide:true];
            [self changeStackViewButton:self.btEnd hide:false];
            [self.lbUsername setHidden:false];
            [self.lbStatus setHidden:false];
            [self.localContainer setHidden:true];
            [self.stackVideoCallAction setHidden:true];
            [self.stackVoiceCallAction setHidden:false];
            break;
        case CallScreenTypeOutgoingVideo:
            [self changeStackViewButton:self.btAccept hide:true];
            [self changeStackViewButton:self.btReject hide:true];
            [self changeStackViewButton:self.btEnd hide:false];
            [self.lbUsername setHidden:false];
            [self.lbStatus setHidden:false];
            [self.localContainer setHidden:true];
            [self.stackVideoCallAction setHidden:true];
            [self.stackVoiceCallAction setHidden:false];
            break;
        case CallScreenTypeIncomingVoice:
            [self changeStackViewButton:self.btAccept hide:false];
            [self changeStackViewButton:self.btReject hide:false];
            [self changeStackViewButton:self.btEnd hide:true];
            [self.lbUsername setHidden:false];
            [self.lbStatus setHidden:true];
            [self.localContainer setHidden:true];
            [self.stackVideoCallAction setHidden:true];
            [self.stackVoiceCallAction setHidden:true];
            break;
        case CallScreenTypeIncomingVideo:
            [self changeStackViewButton:self.btAccept hide:false];
            [self changeStackViewButton:self.btReject hide:false];
            [self changeStackViewButton:self.btEnd hide:true];
            [self.lbUsername setHidden:false];
            [self.lbStatus setHidden:true];
            [self.localContainer setHidden:true];
            [self.stackVideoCallAction setHidden:true];
            [self.stackVoiceCallAction setHidden:true];
            break;
        case CallScreenTypeCallingVoice:
            [self changeStackViewButton:self.btAccept hide:true];
            [self changeStackViewButton:self.btReject hide:true];
            [self changeStackViewButton:self.btEnd hide:false];
            [self.lbUsername setHidden:false];
            [self.lbStatus setHidden:false];
            [self.localContainer setHidden:true];
            [self.stackVideoCallAction setHidden:true];
            [self.stackVoiceCallAction setHidden:false];
            break;
        case CallScreenTypeCallingVideo:
            [self changeStackViewButton:self.btAccept hide:true];
            [self changeStackViewButton:self.btReject hide:true];
            [self changeStackViewButton:self.btEnd hide:true];
            [self.lbUsername setHidden:true];
            [self.lbStatus setHidden:true];
            [self.localContainer setHidden:false];
            [self.stackVideoCallAction setHidden:false];
            [self.stackVoiceCallAction setHidden:true];
            break;
        default:
            break;
    }
    
}

- (CallScreenType)getScreenType {
    CallScreenType screenType;
    if (_signalingState == SignalingStateAnswered) {
        screenType = _isVideo ? CallScreenTypeCallingVideo : CallScreenTypeCallingVoice;
    } else {
        if (_isIncoming) {
            screenType = _isVideo ? CallScreenTypeIncomingVideo : CallScreenTypeIncomingVoice;
        } else {
            screenType = _isVideo ? CallScreenTypeOutgoingVideo : CallScreenTypeOutgoingVoice;
        }
    }
    
    return screenType;
}

//MARK: -  Action

- (IBAction)endCallTapped:(UIButton *)sender {
    [[CallManager instance] hangup:nil];
}

- (IBAction)muteTapped:(UIButton *)sender {
    _isMute = !_isMute;
    [[CallManager instance] mute:_isMute completion:^(BOOL status) {
        NSString *imageName = _isMute ? @"ic-mute-selected-new" : @"ic-mute-new";
        [self.btMute setBackgroundImage:[UIImage imageNamed:imageName] forState:UIControlStateNormal];
    }];
}

- (IBAction)speakerTapped:(UIButton *)sender {
    _isSpeaker = !_isSpeaker;
    [[StringeeAudioManager instance] setLoudspeaker:_isSpeaker];
    NSString *imageName = _isSpeaker ? @"ic-speaker-selected-new" : @"ic-speaker-new";
    [self.btSpeaker setBackgroundImage:[UIImage imageNamed:imageName] forState:UIControlStateNormal];
}


- (IBAction)acceptTapped:(UIButton *)sender {
    if (@available(iOS 14, *)) {
        [[CallManager instance] answerCallkitCall];
    } else {
        [[CallManager instance] answer:true];
    }
}

- (IBAction)rejectTapped:(UIButton *)sender {
    [[CallManager instance] reject:nil];
}

- (IBAction)videoEndTapped:(id)sender {
    [[CallManager instance] hangup:nil];
}

- (IBAction)cameraTapped:(id)sender {
    [self.stringeeCall switchCamera];
}

- (IBAction)videoMuteTapped:(id)sender {
    _isMute = !_isMute;
    [[CallManager instance] mute:_isMute completion:^(BOOL status) {
        NSString *imageName = _isMute ? @"ic-mute-selected-new" : @"ic-mute-new";
        [self.btVideoMute setBackgroundImage:[UIImage imageNamed:imageName] forState:UIControlStateNormal];
    }];
}

- (IBAction)videoTapped:(id)sender {
    _localVideoEnabled = !_localVideoEnabled;
    [self.stringeeCall enableLocalVideo:_localVideoEnabled];
    NSString *imageName = _localVideoEnabled ? @"ic-video-new" : @"ic-video-selected-new";
    [self.btVideo setBackgroundImage:[UIImage imageNamed:imageName] forState:UIControlStateNormal];
}

// MARK: - Timer

- (void)startTimeoutTimer {
    if (timeoutTimer != nil) {
        return;
    }
    
    timeoutTimer = [NSTimer scheduledTimerWithTimeInterval:10 target:self selector:@selector(checkCallTimeOut) userInfo:nil repeats:YES];
    [[NSRunLoop currentRunLoop] addTimer:timeoutTimer forMode:NSDefaultRunLoopMode];
    [timeoutTimer fire];
}

- (void)stopTimeoutTimer {
    CFRunLoopStop(CFRunLoopGetCurrent());
    [timeoutTimer invalidate];
    timeoutTimer = nil;
}
- (void)checkCallTimeOut {
    NSLog(@"checkCallTimeOut");
    interval += 10;
    if (interval >= CALL_TIME_OUT && !timer) {
        if (_isIncoming) {
            [[CallManager instance] reject:nil];
        } else {
            [[CallManager instance] hangup:nil];
        }
    }
}

- (void)startTimer {
    if (_signalingState != SignalingStateAnswered || _mediaState != MediaStateConnected) {
        return;
    }
    
    if (!timer) {
        timer = [NSTimer scheduledTimerWithTimeInterval:1.0 target:self selector:@selector(timerTick:) userInfo:nil repeats:YES];
        [[NSRunLoop currentRunLoop] addTimer:timer forMode:NSDefaultRunLoopMode];
        [timer fire];
    }
}

- (void)timerTick:(NSTimer *)timer {
    self.timeSec++;
    if (self.timeSec == 60)
    {
        self.timeSec = 0;
        self.timeMin++;
    }
    NSString* timeNow = [NSString stringWithFormat:@"%02d:%02d", self.timeMin, self.timeSec];
    if (self.lbStatus.hidden) {
        self.lbStatus.hidden = NO;
    }
    self.lbStatus.text= timeNow;
}

// Kết thúc đếm thời gian cuộc gọi
- (void)stopTimer {
    CFRunLoopStop(CFRunLoopGetCurrent());
    [timer invalidate];
    timer = nil;
    NSString* timeNow = [NSString stringWithFormat:@"%02d:%02d", self.timeMin, self.timeSec];
    self.lbStatus.text= timeNow;
}

// MARK: - StringeeCallDelegate

- (void)didChangeSignalingState:(StringeeCall *)stringeeCall signalingState:(SignalingState)signalingState reason:(NSString *)reason sipCode:(int)sipCode sipReason:(NSString *)sipReason {
    NSLog(@"*********Callstate: %ld", (long)signalingState);
    _signalingState = signalingState;
    dispatch_async(dispatch_get_main_queue(), ^{
        switch (signalingState) {
            case SignalingStateCalling:
                self.lbStatus.text = @"Calling";
                break;
            case SignalingStateRinging:
                self.lbStatus.text = @"Ringing";
                break;
            case SignalingStateAnswered:
                [self updateScreen];
                [self startTimer];
                break;
            case SignalingStateBusy:
                [self endCallAndDismissWithTitle:@"Busy"];
                break;
            case SignalingStateEnded:
                [self endCallAndDismissWithTitle:nil];
                break;
        }
    });
}

- (void)didChangeMediaState:(StringeeCall *)stringeeCall mediaState:(MediaState)mediaState {
    dispatch_async(dispatch_get_main_queue(), ^{
        _mediaState = mediaState;
        switch (mediaState) {
            case MediaStateConnected:
                [self startTimer];
                
                // if call's type is video then route audio to speaker
                if (!_mediaFirstTimeConnected) {
                    _mediaFirstTimeConnected = !_mediaFirstTimeConnected;
                    if (_isVideo) {
                        [self routeToSpeakerIfNeeded];
                    }
                    [self stopTimeoutTimer];
                }
                
                break;
            case MediaStateDisconnected:
                break;
            default:
                break;
        }
    });
}

- (void)didReceiveLocalStream:(StringeeCall *)stringeeCall {
    dispatch_async(dispatch_get_main_queue(), ^{
        stringeeCall.localVideoView.frame = CGRectMake(0, 0, self.localContainer.frame.size.width, self.localContainer.frame.size.height);
        [self.localContainer insertSubview:stringeeCall.localVideoView atIndex:0];
    });
}

- (void)didReceiveRemoteStream:(StringeeCall *)stringeeCall {
    dispatch_async(dispatch_get_main_queue(), ^{
        stringeeCall.remoteVideoView.frame = CGRectMake(0, 0, self.view.frame.size.width, self.view.frame.size.height);
        [self.view insertSubview:stringeeCall.remoteVideoView atIndex:0];
    });
}

- (void)didHandleOnAnotherDevice:(StringeeCall *)stringeeCall signalingState:(SignalingState)signalingState reason:(NSString *)reason sipCode:(int)sipCode sipReason:(NSString *)sipReason {
    NSLog(@"didHandleOnAnotherDevice %ld", (long)signalingState);
    if (signalingState == SignalingStateAnswered || signalingState == SignalingStateBusy || signalingState == SignalingStateEnded) {
        [self endCallAndDismissWithTitle:@"Call is handled on another device"];
    }
}

// MARK: - AudioSession

- (void)handleAudioSessionRouteChange {
    NSLog(@"handleAudioSessionRouteChange");
    dispatch_async(dispatch_get_main_queue(), ^{
        AVAudioSessionRouteDescription *currentRoute = [AVAudioSession sharedInstance].currentRoute;
        AVAudioSessionPortDescription *portDes = currentRoute.outputs.firstObject;
        if (portDes.portType == AVAudioSessionPortBuiltInSpeaker) {
            _isSpeaker = true;
        } else if (portDes.portType == AVAudioSessionPortHeadphones || portDes.portType == AVAudioSessionPortBuiltInReceiver) {
            _isSpeaker = false;
        }
        
        NSString *imageName = _isSpeaker ? @"ic-speaker-selected-new" : @"ic-speaker-new";
        [self.btSpeaker setBackgroundImage:[UIImage imageNamed:imageName] forState:UIControlStateNormal];
    });
}

- (void)routeToSpeakerIfNeeded {
    dispatch_async(dispatch_get_main_queue(), ^{
        AVAudioSessionRouteDescription *currentRoute = [AVAudioSession sharedInstance].currentRoute;
        AVAudioSessionPortDescription *portDes = currentRoute.outputs.firstObject;
        if (portDes.portType != AVAudioSessionPortHeadphones && ![self isBluetoothConnected]) {
            [[StringeeAudioManager instance] setLoudspeaker:true];
            _isSpeaker = true;
        }
    });
}

- (BOOL)isBluetoothConnected {
    NSArray<AVAudioSessionPortDescription *> *availableInputs = [AVAudioSession sharedInstance].availableInputs;
    if (availableInputs == nil || availableInputs.count == 0) {
        return false;
    }
    
    for (AVAudioSessionPortDescription *input in availableInputs) {
        if (input.portType == AVAudioSessionPortBluetoothLE || input.portType == AVAudioSessionPortBluetoothHFP || input.portType == AVAudioSessionPortBluetoothA2DP) {
            return true;
        }
    }
    
    return false;
}


// MARK: - Utils

- (void)delayCallback:(void(^)(void))callback forTotalSeconds:(double)delayInSeconds {
    dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW, delayInSeconds * NSEC_PER_SEC);
    dispatch_after(popTime, dispatch_get_main_queue(), ^(void){
        if(callback){
            callback();
        }
    });
}

- (void)changeStackViewButton:(UIButton *)button hide:(BOOL)hide {
    button.alpha = hide ? 0 : 1;
    button.enabled = !hide;
}

@end

File: CallkitSample-iOS13/CallkitSample-iOS13/StringeeImplement/CallingViewController.swift
//
//  CallingViewController.swift
//  CallkitSample-Swift
//
//  Created by Thịnh Giò on 2/10/20.
//  Copyright © 2020 ThinhNT. All rights reserved.
//

import UIKit

struct CallControl {
    var isIncoming = false
    var isAppToPhone = false
    var isVideo = false
    
    var from = ""
    var to = ""
    var username = ""
    var displayName: String {
        if username.count > 0 {
            return username
        } else {
            return isIncoming ? from : to
        }
    }
    
    var isMute = false
    var isSpeaker = false
    var localVideoEnabled = true
    var signalingState: SignalingState = .calling
    var mediaState: MediaState = .disconnected
}

enum CallScreenType {
    case outgoing
    case incoming
    case calling
}

class TimeCounter {
    var sec: Int = 0
    var min: Int = 0
    var hour: Int = 0
    
    func timeNow() -> String {
        sec = sec + 1
        if sec == 60 {
            sec = 0
            min = min + 1
        }
        
        if min == 60 {
            min = 0
            hour = hour + 1
        }
        
        return currentTime()
    }
    
    func currentTime() -> String {
        if hour > 0 {
            return String(format: "%02d:%02d:%02d", hour, min, sec)
        } else {
            return String(format: "%02d:%02d", min, sec)
        }
    }
    
    func hasStarted() -> Bool {
        if sec != 0 || min != 0 || hour != 0 {
            return true
        }
        
        return false
    }
    
    func reset() {
        sec = 0
        min = 0
        hour = 0
    }
}

class CallingViewController: UIViewController {
    
    @IBOutlet weak var lbName: UILabel!
    @IBOutlet weak var lbStatus: UILabel!
    @IBOutlet weak var ivQuality: UIImageView!
    @IBOutlet weak var btMute: UIButton!
    @IBOutlet weak var btSpeaker: UIButton!
    @IBOutlet weak var btEnd: UIButton!
    @IBOutlet weak var btReject: UIButton!
    @IBOutlet weak var btAnswer: UIButton!
    @IBOutlet weak var blurView: UIView!
    
    var callControl: CallControl!
    var call: StringeeCall!
    
    var callTimer: Timer?
    lazy var timeCounter = TimeCounter()
    
    var timeoutTimer: Timer?
    var callInterval: Int = 0
    
    // MARK: - Init
    init(control: CallControl, call: StringeeCall?) {
        super.init(nibName: "CallingViewController", bundle: nil)
        self.callControl = control
        self.call = call
        call?.delegate = self
        InstanceManager.shared.callingVC = self
        
        // Lưu thông tin vào call control
        if let call = call {
            self.callControl.isIncoming = call.isIncomingCall
            self.callControl.isVideo = call.isVideoCall
            self.callControl.from = call.from
            self.callControl.to = call.to
            self.callControl.username = call.fromAlias
            self.callControl.isAppToPhone = call.callType == .callIn || call.callType == .callOut
        }
        
        // Nếu là videoCall thì cho ra loa ngoài
        if self.callControl.isVideo {
            StringeeAudioManager.instance()?.setLoudspeaker(true)
            self.callControl.isSpeaker = true
        }
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Notifications
        NotificationCenter.default.addObserver(self, selector: #selector(CallingViewController.handleSessionRouteChange), name: AVAudioSession.routeChangeNotification, object: nil)
        
        // UI
        setupUI()
        
        // Check timeout for call
        timeoutTimer = Timer.scheduledTimer(timeInterval: 10, target: self, selector: #selector(CallingViewController.checkCallTimeout), userInfo: nil, repeats: true)
        RunLoop.current.add(timeoutTimer!, forMode: .default)
        
        if call == nil {
            call = StringeeCall(stringeeClient: StringeeImplement.shared.stringeeClient, from: callControl.from, to: callControl.to)
            call.delegate = self
            call.isVideoCall = callControl.isVideo
            
            call.make { [weak self] (status, code, message, data) in
                guard let self = self else { return }
                if (status) {
                    // Thành công => start CallKit
                    CallManager.shared.startCall(phone: self.callControl.to, calleeName: self.callControl.displayName, isVideo: self.callControl.isVideo, stringeCall: self.call)
                } else {
                    // Thất bại -> dismissview
                    self.endCallAndDismis()
                }
            }
        }
    }
    
    // MARK: - Outlet Actions
    
    @IBAction func endTapped(_ sender: Any) {
        CallManager.shared.hangup()
    }
    
    @IBAction func rejectTapped(_ sender: Any) {
        CallManager.shared.reject()
    }
    
    @IBAction func answerTapped(_ sender: Any) {
        if #available(iOS 14, *) {
            CallManager.shared.answerCallkitCall()
        } else {
            CallManager.shared.answer()
        }
    }
    
    @IBAction func muteTapped(_ sender: Any) {
        CallManager.shared.mute()
        let imageName = callControl.isMute ? "icon_mute_selected" : "icon_mute"
        btMute.setBackgroundImage(UIImage(named: imageName), for: .normal)
    }
    
    @IBAction func speakerTapped(_ sender: Any) {
        callControl.isSpeaker = !callControl.isSpeaker
        StringeeAudioManager.instance()?.setLoudspeaker(callControl.isSpeaker)
        let imageName = callControl.isSpeaker ? "icon_speaker_selected" : "icon_speaker"
        btSpeaker.setBackgroundImage(UIImage(named: imageName), for: .normal)
    }
    
    // MARK: - Public Actions
    
    func endCallAndDismis(description: String = "Call ended") {
        DispatchQueue.main.async {
            UIDevice.current.isProximityMonitoringEnabled = false
            UIApplication.shared.isIdleTimerDisabled = false
            self.view.isUserInteractionEnabled = false
            self.lbStatus.text = description
            
            // End callkit
            CallManager.shared.endCall()

            // Ngừng timer
            self.stopCallTimer()
            self.stopTimeoutTimer()
            
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.7, execute: {
                self.dismiss(animated: true, completion: nil)
                InstanceManager.shared.callingVC = nil
            })
        }
    }
    
    func updateScreen() {
        DispatchQueue.main.async {
            let screenType = self.screenType()
            switch screenType {
            case .incoming:
                self.btReject.isHidden = false
                self.btAnswer.isHidden = false
                self.btEnd.isHidden = true
                break
            case .outgoing:
                self.btReject.isHidden = true
                self.btAnswer.isHidden = true
                self.btEnd.isHidden = false
                break
            case .calling:
                self.btReject.isHidden = true
                self.btAnswer.isHidden = true
                self.btEnd.isHidden = false
                break
            }
        }
    }
    // MARK: - Private Actions
    
    private func screenType() -> CallScreenType {
        var screenType: CallScreenType!
        if (callControl.signalingState == .answered) {
            screenType = .calling
        } else {
            screenType = callControl.isIncoming ? .incoming : .outgoing
        }
        
        return screenType
    }
    
    private func setupUI() {
        UIDevice.current.isProximityMonitoringEnabled = true
        UIApplication.shared.isIdleTimerDisabled = callControl.isVideo
        
        self.btAnswer.isEnabled = false
        self.btReject.isEnabled = false
        
        // Fill data
        self.lbStatus.text = callControl.isIncoming ? "Incoming Call" : "Outgoing Call"
        self.lbName.text = callControl.displayName
        
        updateScreen()
    }
    
    // MARK: - Timer
    
    private func startCallTimer() {
        if callControl.signalingState != .answered || callControl.mediaState != .connected {
            return
        }
        
        if callTimer == nil {
            // Bắt đầu đếm giây
            callTimer = Timer(timeInterval: 1, target: self, selector: #selector(CallingViewController.timeTick(timer:)), userInfo: nil, repeats: true)
            RunLoop.current.add(callTimer!, forMode: .default)
            callTimer?.fire()
            
            // => Ko check timeout nữa
            self.stopTimeoutTimer()
        }
    }
    
    @objc private func timeTick(timer: Timer) {
        let timeNow = timeCounter.timeNow()
        self.lbStatus.text = timeNow
    }
    
    private func stopCallTimer() {
        CFRunLoopStop(CFRunLoopGetCurrent())
        callTimer?.invalidate()
        callTimer = nil
    }
    
    private func stopTimeoutTimer() {
        CFRunLoopStop(CFRunLoopGetCurrent())
        timeoutTimer?.invalidate()
        timeoutTimer = nil
    }
    
    @objc private func checkCallTimeout() {
        print("checkCallTimeout")
        callInterval += 10
        if callInterval > 60 && callTimer == nil {
            if callControl.isIncoming {
                CallManager.shared.reject()
            } else {
                CallManager.shared.hangup()
            }
        }
    }
}

extension CallingViewController: StringeeCallDelegate {
    func didChangeSignalingState(_ stringeeCall: StringeeCall!, signalingState: SignalingState, reason: String!, sipCode: Int32, sipReason: String!) {
        print("didChangeSignalingState \(signalingState.rawValue)")
        callControl.signalingState = signalingState
        DispatchQueue.main.async {
            switch signalingState {
            case .calling:
                self.lbStatus.text = "Calling..."
            case .ringing:
                self.lbStatus.text = "Ringing..."
            case .answered:
                self.callControl.signalingState = .answered
                self.updateScreen()
                self.startCallTimer()
            case .ended:
                self.endCallAndDismis()
            case .busy:
                self.endCallAndDismis(description: "Busy")
            @unknown default:
                break
            }
        }
    }
    
    func didChangeMediaState(_ stringeeCall: StringeeCall!, mediaState: MediaState) {
        print("didChangeMediaState \(mediaState.rawValue)")
        DispatchQueue.main.async {
            switch mediaState {
            case .connected:
                self.callControl.mediaState = mediaState
                self.lbStatus.text = self.callControl.isIncoming ? "Incoming Call" : "Outgoing Call"
                self.startCallTimer()
            case .disconnected:
                break
            @unknown default:
                break
            }
        }
    }
    
    func didHandle(onAnotherDevice stringeeCall: StringeeCall!, signalingState: SignalingState, reason: String!, sipCode: Int32, sipReason: String!) {
        if signalingState == .answered {
            endCallAndDismis(description: "The call is handled on another device")
        }
    }
}

// MARK: - Notifications

extension CallingViewController {
    @objc private func handleSessionRouteChange() {
        DispatchQueue.main.async {
            let route = AVAudioSession.sharedInstance().currentRoute
            if let portDes = route.outputs.first {
                self.callControl.isSpeaker = portDes.portType == .builtInSpeaker
                let imageName = self.callControl.isSpeaker ? "icon_speaker_selected" : "icon_speaker"
                self.btSpeaker.setBackgroundImage(UIImage(named: imageName), for: .normal)
            }
        }
    }
}

